src/types.ts

```diff
@@ -0,0 +1,66 @@
+import * as crossFetch from 'cross-fetch'
+import * as nodeFetch from 'node-fetch'
+import { IFetchComponent } from '@well-known-components/interfaces'
+
+export type RequestOptions = nodeFetch.RequestInit & {
+  timeout?: number
+  attempts?: number
+  retryDelay?: number
+}
+
+export function createFetchComponent(defaultHeaders?: Headers): IFetchComponent {
+  async function fetch(url: nodeFetch.RequestInfo, options?: RequestOptions): Promise<nodeFetch.Response> {
+    const { timeout, attempts = 1, retryDelay = 0, ...fetchOptions } = options || {}
+    let timer: NodeJS.Timeout | null = null
+    let response: Response = new Response()
+
+    if (defaultHeaders) {
+      fetchOptions.headers = { ...(fetchOptions.headers || {}), ...defaultHeaders }
+    }
+
+    let currentAttempt = 0
+    do {
+      try {
+        const controller = new AbortController()
+        const { signal: timeoutSignal } = controller
+
+        if (timeout) {
+          timer = setTimeout(() => {
+            controller.abort()
+          }, timeout)
+        }
+
+        const fetchPromise = crossFetch.default(url.toString(), {
+          ...fetchOptions
+        } as any)
+
+        const racePromise = Promise.race([
+          fetchPromise,
+          new Promise((resolve, _) => {
+            timeoutSignal.addEventListener('abort', () => {
+              response = new Response('timeout', { status: 408, statusText: 'Request Timeout' })
+              resolve(response)
+            })
+          })
+        ])
+
+        ++currentAttempt
+        response = (await racePromise) as Response
+
+        if (timer) clearTimeout(timer)
+      } finally {
+        if (response.ok || currentAttempt >= attempts) break
+        else await new Promise((resolve) => setTimeout(resolve, retryDelay))
+      }
+    } while (!response.ok && currentAttempt < attempts)
```

Comment:
Beware that only DELETE & PUT (unsafe) and GET, HEAD & OPTIONS (safe) are automatically retry-able http methods. If we are adding the logic to handle it I'd be more selective to filter the methods and dig deeper about the `response.ok` condition. For example, a 401 is not ok, but it is kind of a response that makes us not want to try again.

POST on the other hand, is never retryi-able

```suggestion
    } while (!response.ok && currentAttempt < attempts)
```

Reply:
Thanks for this feedback. I just updated the way we manage the retry logic, I've considered the following methods as idempotent, thereby will be retried if attempts > 1 is provided: GET, HEAD, OPTIONS, DELETE, PUT.

Also, I've updated the code to avoid retrying if the response status code is any of: 404, 400, 401, 403 (I'm not sure about adding 5XX codes)

If this is ok, I'll start refactoring the implementation and add more use cases to the test suite.

Reply:
I think 5xx 404 403 and 400 are sensible retries. Seems to be a good first approach. The "explicit attempts" is 👌

Do you have some data about which retries are often performed by our servers and why?

Reply:
I'm thinking that I can also make this blacklist dynamic through the component configuration but I'm not sure about complexing the creation of the component. There was also another idea I have been told: dynamically configures whether this component throws when code != 2XX or just returns the proper response object.